### PR TITLE
DNM: netty: add handler to combine small writes into larger buffer

### PR DESCRIPTION
**Pull request by buchgr**
This is a proof of concept that combines smaller `ByteBufs` into a larger `ByteBuf` by copying. The idea came out of a discussion with @normanmaurer today, where he dropped his wisdom on me :-). Amongst others he mentioned that passing many small buffers to `writev` can often be much slower than one big buffer to `write`. I decided to give it a try and implemented a `WriteCombiningHandler` that simply copies many smaller buffers into a bigger buffer. See the benchmark below.

Now, this is still quite early and I have to experiment and benchmark a lot more. Also, this handler does not interact well with a channel's writability and promise's aren't implemented correctly (yet). Also, the handler is quite dumb in that it just combines all writes. We could instead only combine buffers less than N bytes, and have different sized buffers to optimize for memory usage (i.e. don't allocate a 4KB buffer to merge 5 writes totalling 250 bytes). Many more sophisticated heuristics are conceivable.

Furthermore, @normanmaurer mentioned that if real performance benefits can be shown, this capability should be added to Netty's `ChannelOutboundBuffer`.

```
./qps_client --client_payload=20 --server_payload=40 --outstanding_rpcs=2000 
--channels=1 --streaming_rpcs --directexecutor --address=localhost:8080
```

```
./qps_server --directexecutor --address=localhost:8080
```

Best out of 3 runs:

Before:
```
Channels:                       1
Outstanding RPCs per Channel:   2000
Server Payload Size:            40
Client Payload Size:            20
50%ile Latency (in micros):     10367
90%ile Latency (in micros):     13743
95%ile Latency (in micros):     14703
99%ile Latency (in micros):     18255
99.9%ile Latency (in micros):   36159
Maximum Latency (in micros):    129471
QPS:                            189670
```

After:
```
Channels:                       1
Outstanding RPCs per Channel:   2000
Server Payload Size:            40
Client Payload Size:            20
50%ile Latency (in micros):     9007
90%ile Latency (in micros):     10799
95%ile Latency (in micros):     11895
99%ile Latency (in micros):     13495
99.9%ile Latency (in micros):   22431
Maximum Latency (in micros):    107327
QPS:                            217794
```

Thoughts? @ejona86 @nmittler @carl-mastrangelo @louiscryan

### CODE REVIEW

**Comment by normanmaurer** [comment]
You also need to ensure you release / flush the buffer when the channel is closed or the handler is removed

File: netty/src/main/java/io/grpc/netty/WriteCombiningHandler.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufAllocator;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelOutboundHandlerAdapter;
+import io.netty.channel.ChannelPromise;
+import io.netty.util.ReferenceCountUtil;
+
+public class WriteCombiningHandler extends ChannelOutboundHandlerAdapter {
+
+  private static final int BUFFER_SIZE = 4096;
```
Comment (carl-mastrangelo) [inline netty/src/main/java/io/grpc/netty/WriteCombiningHandler.java]: Magic numbers deserve documentation.   In fact, you may want to try testing with different values.   8k would be interesting, as well as powers of two down to maybe 512.

File: netty/src/main/java/io/grpc/netty/WriteCombiningHandler.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+  @Override
+  public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) {
+    ByteBuf data = (ByteBuf) msg;
+    if (buffer == null) {
+      buffer = allocator.directBuffer(BUFFER_SIZE, BUFFER_SIZE);
+    }
+
+    if (buffer.writableBytes() >= data.readableBytes()) {
+      buffer.writeBytes(data);
+      promise.setSuccess();
```
Comment (carl-mastrangelo) [inline netty/src/main/java/io/grpc/netty/WriteCombiningHandler.java]: When I was trying to implement this, I had a very hard time with the promises.  Here, don't you need to wait to fullfill the promise when you flush the large buffer?  I.e. chain the promises?

File: netty/src/main/java/io/grpc/netty/WriteCombiningHandler.java (hunk 1 above)
Comment (carl-mastrangelo) [inline netty/src/main/java/io/grpc/netty/WriteCombiningHandler.java]: A common pattern in gRPC is the following write pattern:

little, little, little, BIG.

A larger buffer doesn't help a lot in that case, since you are going to flush pretty much immediately after the the first few writes.  For example, connection initialization and headers would be the small buffers, and the payload of the RPC the big one.

File: netty/src/main/java/io/grpc/netty/WriteCombiningHandler.java (hunk 2 above)
Comment (buchgr) [inline netty/src/main/java/io/grpc/netty/WriteCombiningHandler.java]: @carl-mastrangelo this is just a PoC, that is demonstrating that copying small buffers into a big buffer can provide a speed up (as mentioned in the first message of the PR). It's not correct.

File: netty/src/main/java/io/grpc/netty/WriteCombiningHandler.java (hunk 1 above)
Comment (buchgr) [inline netty/src/main/java/io/grpc/netty/WriteCombiningHandler.java]: @carl-mastrangelo yes this is just a PoC for this particular benchmark where I know that we only write small buffers. As mentioned in the PR message, we will need better heuristics.

File: netty/src/main/java/io/grpc/netty/WriteCombiningHandler.java (hunk 2 above)
Comment (carl-mastrangelo) [inline netty/src/main/java/io/grpc/netty/WriteCombiningHandler.java]: I bring it up because that's where the complexity lies.  This POC is good, and it shows promise.  The difficulty though is a little bit further down the road.

File: netty/src/main/java/io/grpc/netty/WriteCombiningHandler.java (hunk 1 above)
Comment (buchgr) [inline netty/src/main/java/io/grpc/netty/WriteCombiningHandler.java]: @carl-mastrangelo I tried several different sizes. 4K seems to be close to ideal for this benchmark :-).

Comment (carl-mastrangelo) [inline netty/src/main/java/io/grpc/netty/WriteCombiningHandler.java]: Try it with a larger payload too?  I would imagine a 4k payload would make for a more interesting Benchmark.

Comment (buchgr) [inline netty/src/main/java/io/grpc/netty/WriteCombiningHandler.java]: @carl-mastrangelo yes! My plan is to now experiment with the `write` and `writev` system call directly in a little C program and see how they behave.

File: netty/src/main/java/io/grpc/netty/WriteCombiningHandler.java (hunk 2 above)
Comment (nmittler) [inline netty/src/main/java/io/grpc/netty/WriteCombiningHandler.java]: FYI, I think the [SimpleChannelPromiseAggregator](https://github.com/netty/netty/blob/4.1/codec-http2/src/main/java/io/netty/handler/codec/http2/Http2CodecUtil.java#L218) is probably the way to go here. Perhaps we should make that a top-level class in Netty and get it out of the HTTP2 package.

File: netty/src/main/java/io/grpc/netty/WriteCombiningHandler.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+      } else {
+        ctx.write(msg, promise);
+      }
+    }
+  }
+
+  @Override
+  public void flush(ChannelHandlerContext ctx) throws Exception {
+    if (buffer != null && buffer.readableBytes() > 0) {
+      ctx.write(buffer, ctx.voidPromise());
```
Comment (nmittler) [inline netty/src/main/java/io/grpc/netty/WriteCombiningHandler.java]: Maybe consider slicing the buffer if we still have capacity?

File: netty/src/main/java/io/grpc/netty/WriteCombiningHandler.java (hunk 2 above)
Comment (buchgr) [inline netty/src/main/java/io/grpc/netty/WriteCombiningHandler.java]: @nmittler I saw that, but this situation is a bit different. 1) We need to maintain the actual promises and not just a count 2) We need to complete the promises as their respective buffers are written to the network. That is, we need a mapping of byte ranges to promises (promise 1 completed once bytes [0,100] have been written, promise 2 once bytes [101, 150] have been written and so on.). Now we sure don't need to complete each promise immediately once its corresponding bytes have been flushed, but we still need to handle the case where only a fraction of the aggregated ByteBuf has been flushed. Good thing Netty already comes with this capability, namely [ChannelProgressivePromise](https://github.com/netty/netty/blob/4.1/transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java#L235). So I think this makes this quite straightfoward.

Comment (buchgr) [inline netty/src/main/java/io/grpc/netty/WriteCombiningHandler.java]: After a video chat discussion with @nmittler, we concluded that it's easiest to initially implement this with Netty's `PromiseCombiner`. Eventually, we might choose implement it with a special `ChannelProgressivePromise` aggregator.

**Comment by buchgr** [comment]
I am currently busy with exams :( ... there is some thing happening in Netty land though. See https://github.com/netty/netty/pull/5510 and https://github.com/netty/netty/pull/5516

**Comment by buchgr** [comment]
Some thoughts so that I don't forget. 

- Allocate one large buffer (say 16K) and copy small buffers into it. Then slice it, write it and reuse the remaining buffer. We want to amortize allocations across many flushes.
- When should we stop copying? If we encounter a buffer larger than N bytes (to be determined). 
- Maybe we should only copy if there are at least M buffers less than N bytes? We might want to keep a running history of the last M buffers and then decide.
- When to flush? If `.flush()` gets called or we accumulated enough bytes so that when adding more, the channel's writability would change. We want to avoid triggering a writability change and flush before. We can check that via `Channel.unsafe().outboundBuffer().bytesBeforeUnwritable()`.

**Comment by ejona86** [comment]
We may want the handler to release all buffers when it receives a flush(), since retaining a 16K buffer per idle connection seems a bit much.

**Comment by buchgr** [comment]
@louiscryan @ejona86 @nmittler @normanmaurer 

I am having some trouble designing the benchmark with JMH.

I want to measure `write + flush` performance in Netty. That is, do a bunch of writes and then flush. See if combining those writes to a single write and then flush is faster. Ideally, I would want to get something like like `mbit/s`. My first idea was to continuously enqueue a task on the eventloop that performs the `writes + flush` and via a future listener increments a counter by the flushed bytes. However, this approach is flawed in that it can't deal with writes failing (i.e. send buffer being full). When the send buffer is full, then writes will fail and be queued up in the outbound buffer. So essentially the faster we are writing the more likely that writes will.

My idea to circumvent this problem was to instead measure the duration of each sequence of `writes + flush` and store it in a histogram. I could then compute some kind of mean `write + flush` duration, by ignoring everything above the 90%tile or a few standard deviations from the median or so, effectively ignoring those writes that failed due to e.g. a full send buffer. I implemented this via aux counters in JMH, but that doesn't work cause JMH normalizes the aux counter values to `ops/sec` ... in this benchmark `ops/sec` is meaningless though. The benchmark method is just used to trigger the writes and then waits until all writes are complete (similar as done in the other gRPC JMH methods). Even if it worked with JMH, it's just really complex to interpret the result and I would rather have a single throughput figure instead of a histogram.

Another idea is to have two eventloops. One for writing and the other for reading. Set the I/O ratio of the write eventloop to 100 and then write just as many (or fewer) bytes as the send buffer can hold. Measure the the time it took from writing the first buffer to the last and from that compute a throughput value (i.e. 10MB in 200ms => 50MB/s). However, that also seems a bit hacked.

Any thoughts on how to implement a benchmark properly measuring throughput of channel writes + flush?

**Comment by buchgr** [comment]
TL;DR: In selected QPS benchmarks the write combining is 13%, 5%, 10% faster.

**Current master**
```
Jakobs-MacBook-Pro-4:bin buchgr$ ./qps_client --client_payload=20 --server_payload=40 --outstanding_rpcs=100 --channels=1 --directexecutor --address=localhost:8080
Channels:                       1
Outstanding RPCs per Channel:   100
Server Payload Size:            40
Client Payload Size:            20
50%ile Latency (in micros):     1101
90%ile Latency (in micros):     1279
95%ile Latency (in micros):     1357
99%ile Latency (in micros):     1764
99.9%ile Latency (in micros):   2761
Maximum Latency (in micros):    22879
QPS:                            89280
```

```
Jakobs-MacBook-Pro-4:bin buchgr$ ./qps_client --client_payload=300 --server_payload=600 --outstanding_rpcs=100 --channels=1 --directexecutor --address=localhost:8080
Channels:                       1
Outstanding RPCs per Channel:   100
Server Payload Size:            600
Client Payload Size:            300
50%ile Latency (in micros):     1153
90%ile Latency (in micros):     1304
95%ile Latency (in micros):     1378
99%ile Latency (in micros):     1934
99.9%ile Latency (in micros):   2867
Maximum Latency (in micros):    16559
QPS:                            86529

// Streaming RPC
Jakobs-MacBook-Pro-4:bin buchgr$ ./qps_client --client_payload=20 --server_payload=40 --outstanding_rpcs=2000 --channels=1 --streaming_rpcs --directexecutor --address=localhost:8080
Channels:                       1
Outstanding RPCs per Channel:   2000
Server Payload Size:            40
Client Payload Size:            20
50%ile Latency (in micros):     10431
90%ile Latency (in micros):     12055
95%ile Latency (in micros):     12791
99%ile Latency (in micros):     14831
99.9%ile Latency (in micros):   24927
Maximum Latency (in micros):    99967
QPS:                            188915
```

**With this change**
```
Jakobs-MBP-4:bin buchgr$ ./qps_client --client_payload=20 --server_payload=40 --outstanding_rpcs=100 --channels=1 --directexecutor --address=localhost:8080
Channels:                       1
Outstanding RPCs per Channel:   100
Server Payload Size:            40
Client Payload Size:            20
50%ile Latency (in micros):     948
90%ile Latency (in micros):     1115
95%ile Latency (in micros):     1197
99%ile Latency (in micros):     1519
99.9%ile Latency (in micros):   2349
Maximum Latency (in micros):    38271
QPS:                            101665

Jakobs-MBP-4:bin buchgr$ ./qps_client --client_payload=300 --server_payload=600 --outstanding_rpcs=100 --channels=1 --directexecutor --address=localhost:8080
Channels:                       1
Outstanding RPCs per Channel:   100
Server Payload Size:            600
Client Payload Size:            300
50%ile Latency (in micros):     1063
90%ile Latency (in micros):     1246
95%ile Latency (in micros):     1355
99%ile Latency (in micros):     1704
99.9%ile Latency (in micros):   2691
Maximum Latency (in micros):    15863
QPS:                            90682

// Streaming RPC
Jakobs-MBP-4:bin buchgr$ ./qps_client --client_payload=20 --server_payload=40 --outstanding_rpcs=2000 --channels=1 --streaming_rpcs --directexecutor --address=localhost:8080
Channels:                       1
Outstanding RPCs per Channel:   2000
Server Payload Size:            40
Client Payload Size:            20
50%ile Latency (in micros):     9495
90%ile Latency (in micros):     11231
95%ile Latency (in micros):     12135
99%ile Latency (in micros):     13807
99.9%ile Latency (in micros):   19631
Maximum Latency (in micros):    85119
QPS:                            206568
```

**Comment by carl-mastrangelo** [comment]
@buchgr Those numbers look pretty good.  Could you also run the FlowControlledMessagesPerSecondBenchmark  test before and after?  preferably with a large (10?) number of forks.

File: netty/src/main/java/io/grpc/netty/WriteCombiningHandler.java [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+import io.netty.channel.ChannelOutboundHandlerAdapter;
+import io.netty.channel.ChannelPromise;
+import io.netty.channel.DefaultChannelPromise;
+import io.netty.util.concurrent.Future;
+import io.netty.util.concurrent.FutureListener;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class WriteCombiningHandler extends ChannelOutboundHandlerAdapter {
```
Comment (normanmaurer) [inline netty/src/main/java/io/grpc/netty/WriteCombiningHandler.java]: @buchgr do you plan to push this to netty ?

File: netty/src/main/java/io/grpc/netty/WriteCombiningHandler.java [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+
+  private static final int INITIAL_BUFFER_SIZE = 4096;
+  private static final int MAX_COPY_BUFFER_SIZE = 256;
```
Comment (normanmaurer) [inline netty/src/main/java/io/grpc/netty/WriteCombiningHandler.java]: make these configurable via constructor ?

File: netty/src/main/java/io/grpc/netty/WriteCombiningHandler.java [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+    if (data.readableBytes() > MAX_COPY_BUFFER_SIZE) {
+      writeBuffer();
+      ctx.write(data, promise);
+    } else if (data.readableBytes() > buffer.writableBytes()) {
+      writeBuffer();
+      copyBytes(data, promise);
+    } else {
+      copyBytes(data, promise);
+    }
+  }
```
Comment (normanmaurer) [inline netty/src/main/java/io/grpc/netty/WriteCombiningHandler.java]: @buchgr I think you also want to override `close(...)` and `disconnect(...)` and ensure that the buffer is written before you process these.

File: netty/src/main/java/io/grpc/netty/WriteCombiningHandler.java [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+
+  @Override
+  public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+    buffer = ctx.alloc().directBuffer(INITIAL_BUFFER_SIZE, INITIAL_BUFFER_SIZE);
+    bufferPromise = new CollectivePromise(ctx);
+    this.ctx = ctx;
+  }
+
+  @Override
+  public void handlerRemoved(ChannelHandlerContext ctx) throws Exception {
```
Comment (normanmaurer) [inline netty/src/main/java/io/grpc/netty/WriteCombiningHandler.java]: Maybe try to write the buffer and not just release it ?

File: netty/src/main/java/io/grpc/netty/WriteCombiningHandler.java [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+import java.util.List;
+
+public class WriteCombiningHandler extends ChannelOutboundHandlerAdapter {
+
+  private static final int INITIAL_BUFFER_SIZE = 4096;
+  private static final int MAX_COPY_BUFFER_SIZE = 256;
+
+  private ByteBuf buffer;
+  private CollectivePromise bufferPromise;
+
```
Comment (normanmaurer) [inline netty/src/main/java/io/grpc/netty/WriteCombiningHandler.java]: remove empty line

File: netty/src/main/java/io/grpc/netty/WriteCombiningHandler.java [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+  }
+
+  private void copyBytes(ByteBuf buf, ChannelPromise promise) {
+    buffer.writeBytes(buf);
+    bufferPromise.add(promise);
+    safeRelease(buf);
+  }
+
+  private static final class CollectivePromise extends DefaultChannelPromise
+      implements FutureListener<Void> {
```
Comment (normanmaurer) [inline netty/src/main/java/io/grpc/netty/WriteCombiningHandler.java]: ChannelFutureListener ?

File: netty/src/main/java/io/grpc/netty/WriteCombiningHandler.java [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+
+    @Override
+    public void operationComplete(Future<Void> future) throws Exception {
+      final boolean isSuccess = future.isSuccess();
+      for (int i = 0; i < promises.size(); i++) {
+        ChannelPromise promise = promises.get(i);
+        if (isSuccess) {
+          promise.trySuccess();
+        } else {
+          promise.tryFailure(future.cause());
```
Comment (normanmaurer) [inline netty/src/main/java/io/grpc/netty/WriteCombiningHandler.java]: Store the Throwable to reduce volatile reds ?

File: netty/src/main/java/io/grpc/netty/WriteCombiningHandler.java (hunk 4 above)
Comment (buchgr) [inline netty/src/main/java/io/grpc/netty/WriteCombiningHandler.java]: @normanmaurer yeah sure ... once I get it right :-).

**Comment by buchgr** [comment]
Opened a clean PR #2146
